pontoon/sync/core.py

```diff
@@ -241,3 +252,74 @@ def commit_changes(db_project, vcs_project, changeset, locale):
     locale_path = vcs_project.locale_directory_paths[locale.code]
     repo = db_project.repository_for_path(locale_path)
     repo.commit(commit_message, commit_author, locale_path)
+
+
+def get_changed_locales(db_project, locales):
+    """
+    Narrow down locales to the ones that have changed since the last sync by fetching latest
+    repository commit hashes via API. For projects with many repositories, this is much faster
+    than running VCS pull/clone for each repository.
+    """
+    repos = db_project.translation_repositories()
+
+    # Requirement: all translation repositories must have API configured.
+    for repo in repos:
+        if not (repo.api_endpoint and repo.latest_commit_path):
+            return None
+
+    log.info('Fetching latest commit hashes for project {0} started.'.format(db_project.slug))
+
+    changed_locale_pks = []
+    unchanged_locale_pks = []
+    error_locale_pks = []
+
+    changed_locales_in_db = locales.filter(
+        changedentitylocale__entity__resource__project=db_project
+    )
+
+    for repo in repos:
+        for locale in locales:
+            # If we already processed the locale, we can move on.
+            if locale.pk in changed_locale_pks + unchanged_locale_pks:
+                continue
+
+            # If locale has changed in the DB, we need to sync it.
+            if locale in changed_locales_in_db:
+                changed_locale_pks.append(locale.pk)
+                continue
+
+            try:
+                locale_api_endpoint = repo.api_endpoint.format(locale_code=locale.code)
+                response = requests.get(locale_api_endpoint)
+
+                # Raise exception on 4XX client error or 5XX server error response
+                if response.status_code != 200:
+                    response.raise_for_status()
+
+                # If locale has changed in the VCS, we need to sync it.
+                latest_commit_id = get_json_value_by_path(response.json(), repo.latest_commit_path)
```

Comment:
I wonder if there's a better way of solving this problem.

Currently we have two repository sources for Firefox (and a few other projects):
* https://api.bitbucket.org/2.0/repositories/mozilla-l10n/arn/commits
* https://hg.mozilla.org/l10n-central/ja/rev/default?style=json

Since we need to retrieve commit hash from 2 different JSON files, I added this `latest_commit_path` which tries to simulate what XPath does.